src/poetry/puzzle/provider.py

```diff
@@ -508,10 +511,10 @@ def incompatibilities_for(
         ]
 
     def complete_package(self, package: DependencyPackage) -> DependencyPackage:
-        if package.is_root():
+        if package.package.is_root():
```

Comment:
Let's do the same `package = dependency_package.package` clarification here, like we did above, instead of typing `package.package` everywhere.

Reply:
it gets a lot more awkward in this one, because we create new `package = ...` in various branches and I was afraid of messing it up

Reply:
I'd probably call the root one `package =` then and use a new name in each branch for clarity.